scapy/contrib/mqttsn.py

```diff
@@ -0,0 +1,467 @@
+        ConditionalField(StrFixedLenField("short_topic", None, length=2),
+                         lambda pkt: pkt.tid_type == 0b10),
+        ConditionalField(StrLenField("topic_name", None,
+                                     length_from=_mqttsn_length_from(3)),
+                         lambda pkt: pkt.tid_type not in [0b01, 0b10]),
+    ]
+
+
+class MQTTSNSuback(Packet):
+    name = "MQTT-SN subscribe ACK"
+    fields_desc = FLAG_FIELDS + [
+        ShortField("tid", 0),
+        ShortField("mid", 0),
+        ByteEnumField("return_code", ACCEPTED, RETURN_CODES),
+    ]
+
+
+class MQTTSNUnsubscribe(Packet):
+    name = "MQTT-SN unsubscribe request"
+    fields_desc = FLAG_FIELDS + [
+        ShortField("mid", 0),
+        ConditionalField(ShortField("tid", None),
+                         lambda pkt: pkt.tid_type == 0b01),
+        ConditionalField(StrFixedLenField("short_topic", None, length=2),
+                         lambda pkt: pkt.tid_type == 0b10),
+        ConditionalField(StrLenField("topic_name", None,
+                                     length_from=_mqttsn_length_from(3)),
+                         lambda pkt: pkt.tid_type not in [0b01, 0b10]),
+    ]
+
+
+class MQTTSNUnsuback(Packet):
+    name = "MQTT-SN unsubscribe ACK"
+    fields_desc = [
+        ShortField("mid", 0),
+    ]
+
+
+class MQTTSNPingReq(Packet):
+    name = "MQTT-SN ping request"
+    fields_desc = [
+        StrLenField("client_id", "", length_from=_mqttsn_length_from(0)),
+    ]
+
+
+class MQTTSNPingResp(Packet):
+    name = "MQTT-SN ping response"
+
+
+class MQTTSNDisconnect(Packet):
+    name = "MQTT-SN disconnect request"
+    fields_desc = [
+        ConditionalField(
+            ShortField("duration", None),
+            lambda pkt: hasattr(pkt.underlayer, "len") and
+            ((pkt.underlayer.len is None) or (pkt.underlayer.len > 2))
+        ),
+    ]
+
+
+class MQTTSNWillTopicUpd(Packet):
+    name = "MQTT-SN will topic update"
+    fields_desc = FLAG_FIELDS + [
+        StrLenField("will_topic", "", length_from=_mqttsn_length_from(1)),
+    ]
+
+
+class MQTTSNWillTopicResp(Packet):
+    name = "MQTT-SN will topic response"
+    fields_desc = [
+        ByteEnumField("return_code", ACCEPTED, RETURN_CODES),
+    ]
+
+
+class MQTTSNWillMsgUpd(Packet):
+    name = "MQTT-SN will message update"
+    fields_desc = [
+        StrLenField("will_msg", "", length_from=_mqttsn_length_from(0))
+    ]
+
+
+class MQTTSNWillMsgResp(Packet):
+    name = "MQTT-SN will message response"
+    fields_desc = [
+        ByteEnumField("return_code", ACCEPTED, RETURN_CODES),
+    ]
+
+
+class MQTTSNEncaps(Packet):
+    name = "MQTT-SN encapsulated message"
+    fields_desc = [
+        BitField("resvd", 0, 6),
+        BitField("radius", 0, 2),
+        StrLenField(
+            "w_node_id", "",
+            length_from=_mqttsn_length_from(1)
+        ),
+    ]
+
+
+# Layer bindings
+bind_layers(UDP, MQTTSN, sport=1883)
+bind_layers(UDP, MQTTSN, dport=1883)
```

Comment:
This is incorrect. Will provide some details really soon

Reply:
Done and also provided a test for that case.